pageserver/src/tenant/storage_layer/delta_layer.rs

```diff
@@ -1073,11 +1073,191 @@ impl DeltaLayerInner {
+        let tree_reader = DiskBtreeReader::<_, DELTA_KEY_SIZE>::new(
+            self.index_start_blk,
+            self.index_root_blk,
+            block_reader,
+        );
+
+        let stream = self.stream_index_forwards(&tree_reader, &[0u8; DELTA_KEY_SIZE], ctx);
+        let stream = stream.map_ok(|(key, lsn, pos)| Item::Actual(key, lsn, pos));
+        // put in a sentinel value for getting the end offset for last item, and not having to
+        // repeat the whole read part
+        let stream = stream.chain(futures::stream::once(futures::future::ready(Ok(
+            Item::Sentinel,
+        ))));
+        let mut stream = std::pin::pin!(stream);
+
+        let mut prev: Option<(Key, Lsn, BlobRef)> = None;
+
+        let mut read_builder: Option<VectoredReadBuilder> = None;
+
+        let max_read_size = self
+            .max_vectored_read_bytes
+            .map(|x| x.0.get())
+            .unwrap_or(8192);
+
+        let mut buffer = Some(BytesMut::with_capacity(max_read_size));
+
+        // FIXME: buffering of DeltaLayerWriter
+        let mut per_blob_copy = Vec::new();
+
+        while let Some(item) = stream.try_next().await? {
+            tracing::debug!(?item, "popped");
+            let offset = item
+                .offset()
+                .unwrap_or(BlobRef::new(self.index_start_offset(), false));
+
+            let actionable = if let Some((key, lsn, start_offset)) = prev.take() {
+                let end_offset = offset;
+
+                Some((BlobMeta { key, lsn }, start_offset..end_offset))
+            } else {
+                None
+            };
+
+            let is_last = item.is_last();
+
+            prev = Option::from(item);
+
+            let actionable = actionable.filter(|x| x.0.lsn < truncate_at);
+
+            let builder = if let Some((meta, offsets)) = actionable {
+                // extend or create a new builder
+                if read_builder
+                    .as_mut()
+                    .map(|x| x.extend(offsets.start.pos(), offsets.end.pos(), meta))
+                    .unwrap_or(VectoredReadExtended::No)
+                    == VectoredReadExtended::Yes
+                {
+                    None
+                } else {
+                    read_builder.replace(VectoredReadBuilder::new(
+                        offsets.start.pos(),
+                        offsets.end.pos(),
+                        meta,
+                        max_read_size,
+                    ))
+                }
+            } else {
+                // nothing to do, except perhaps flush any existing for the last element
+                None
+            };
+
+            // flush the possible older builder and also the new one if the item was the last one
+            let builders = builder.into_iter();
+            let builders = if is_last {
+                builders.chain(read_builder.take())
+            } else {
+                builders.chain(None)
+            };
```

Comment:
The sketchiest part is here; you solved this kind of hurdles by a "full" preparation phase (collecting a `Vec<VectoredRead>`), but here we are doing more of an "online" work.

This is not horrible because there's the assertion after the loop.

Reply:
Why can't you use the `VectoredReadPlanner` instead?

I thought it would look something like this (pseuducode):
```
let planner = Vectored::ReadPlanner(your_max_read_size);
for (key, lsn, blob_ref) in stream {
  let flag = if lsn >= truncate_lsn {
    BlobFlag::Ignore
  } else {
    BlobFlag::None
  };
}

let data_end_offset = self.index_start_blk as u64 * PAGE_SZ as u64;
planner.handle(data_end_offset);
let reads = planner.finish();

[...]
// do reads
// pipe results into writer
```

Reply:
Because internally VectoredReadPlanner has a BTreeMap `key => [(lsn, file_offsets)]` which is not needed here. A solution using that would be:

1. scan the whole file
2. do vectored reads
    1. write the result per value

Instead this implementation tries to be more "fill one read up, write results out". I cannot see what the benefit would be with VectoredReadPlanner.

Reply:
The business logic here is the one `Option::filter`.

Instead of trying to solve the pain of needing to do the trailing by filling up a possibly a huge map of all of the delta layer keys, keep those in memory, then read them in the correct order, I read one-by-one since the file is already sorted.

I'd rather solve for the trailing issue by adding another stream wrapper which would give a single pass feel for the run. Even still, I'd need the sentinel value there.